scripts/apiserver/resources.tmpl

```diff
@@ -99,4 +77,87 @@ func (res *{{.res.kind}}) AsInstance() (*apiv1.ResourceInstance, error) {
 	}
 
 	return &instance, nil
-}
\ No newline at end of file
+}
+
+// FromInstance converts a ResourceInstance to a {{.res.kind}}
+func (res *{{.res.kind}}) FromInstance(ri *apiv1.ResourceInstance) (error) {
+    if ri == nil {
+        res = nil
+        return nil
+    }
+    var err error
+    rawResource := ri.GetRawResource()
+    if rawResource == nil {
+        rawResource, err = json.Marshal(ri)
+        if err != nil {
+            return err
+        }
+    }
+    err = json.Unmarshal(rawResource, res)
+    return err
+}
+
+// MarshalJSON custom marshaller to handle sub resources
+func (res *{{.res.kind}}) MarshalJSON() ([]byte, error) {
+	m, err := json.Marshal(&res.ResourceMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	var out map[string]interface{}
+	err = json.Unmarshal(m, &out)
+	if err != nil {
+		return nil, err
+	}
+
+	{{range $field, $isNotEmpty := .res.fields -}}
+        {{- $fieldName := $field | strings.Title -}}
+		{{- $keyName := $field | strings.ToLower -}}
+		out["{{$keyName}}"] = res.{{$fieldName}}
+	{{end}}
+
+	return json.Marshal(out)
+}
+
+// UnmarshalJSON custom unmarshaler to handle sub resources
+func (res *{{.res.kind}}) UnmarshalJSON(data []byte) error {
+    var err error
+
+    aux := &apiv1.ResourceInstance{}
+    err = json.Unmarshal(data, aux)
+    if err != nil {
+        return err
+    }
+
+    res.ResourceMeta = aux.ResourceMeta
+    res.Owner = aux.Owner
+
+    sr, err := json.Marshal(aux.Spec)
+    if err != nil {
+        return err
+    }
+
+    err = json.Unmarshal(sr, &res.Spec)
+    if err != nil {
+        return err
+    }
+
+    {{range $field, $isNotEmpty := .res.fields -}}
+        {{- $fieldName := $field | strings.Title -}}
+        {{- $keyName := $field | strings.ToLower -}}
+        {{if eq $keyName "owner"}}
+        {{else if eq $keyName "spec"}}
+        {{else}}
+        sr, err = json.Marshal(aux.SubResources["{{$keyName}}"])
```

Comment:
could we throw a comment here to note what sub resource is being handled?

// Unmarshalling subresource {{$fieldName}}

....or similar?

Maybe one above the spec handling as well?

Reply:
Good idea

Reply:
isn't the action unmarshaling?

Reply:
that step is marshal, but overall we want to unmarshal the sub resource

Reply:
I guess it is not necessary, but should we regen?